README
Driver for the PCD8544 LCD controller, which is used in the Nokia 5110 display.


CHANGELOG
Revision history for perl module Device::PCD8544

0.481538334543082 2015-01-??

- Image conversion implemented

0.0268329147520525 2015-01-03

- Initial release


examples/rpi.pl
#!perl
# Copyright (c) 2015  Timm Murray
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are met:
# 
#     * Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright 
#       notice, this list of conditions and the following disclaimer in the 
#       documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
#
# Setup using the Sparkfun breakout board:
#
#     LCD Pin         Rpi Pin
#     ========        ================
#     (VCC)  1        GPIO2
#     (GND)  2        GND
#     (SCE)  3        CE0 (GPIO8)
#     (RST)  4        GPIO24
#     (DC)   5        GPIO23
#     (MOSI) 6        MOSI (SPI)
#     (SCLK) 7        CLK (SPI)
#     (LED)  8        3.3V
# 
use v5.14;
use Device::PCD8544;
use Device::WebIO::RaspberryPi;

my @PIC_OUT = (
    0xFF, 0x8D, 0x9F, 0x13, 0x13, 0xF3, 0x01, 0x01, 0xF9, 0xF9, 0x01, 0x81, 0xF9,
    0xF9, 0x01, 0xF1, 0xF9, 0x09, 0x09, 0xFF, 0xFF, 0xF1, 0xF9, 0x09, 0x09, 0xF9,
    0xF1, 0x01, 0x01, 0x01, 0x01, 0x01, 0xF9, 0xF9, 0x09, 0xF9, 0x09, 0xF9, 0xF1,
    0x01, 0xC1, 0xE9, 0x29, 0x29, 0xF9, 0xF1, 0x01, 0xFF, 0xFF, 0x71, 0xD9, 0x01,
    0x01, 0xF1, 0xF9, 0x29, 0x29, 0xB9, 0xB1, 0x01, 0x01, 0x01, 0xF1, 0xF1, 0x11,
    0xF1, 0xF1, 0xF1, 0xE1, 0x01, 0xE1, 0xF1, 0x51, 0x51, 0x71, 0x61, 0x01, 0x01,
    0xC1, 0xF1, 0x31, 0x31, 0xF1, 0xFF, 0xFF, 0x00, 0x01, 0x01, 0x01, 0x01, 0x60,
    0xE0, 0xA0, 0x01, 0x01, 0x81, 0xE1, 0x61, 0x60, 0xC0, 0x01, 0xE1, 0xE1, 0x21,
    0x21, 0xE0, 0xC1, 0x01, 0xC1, 0xE1, 0x20, 0x20, 0xFC, 0xFC, 0xE0, 0xE0, 0xC1,
    0xE1, 0xE0, 0xC1, 0xE0, 0xE1, 0x01, 0xFC, 0xFC, 0x21, 0x21, 0xE1, 0xC1, 0xE5,
    0xE4, 0x01, 0xC1, 0xE0, 0x20, 0x21, 0x20, 0x00, 0x01, 0xFD, 0xFD, 0x21, 0x20,
    0xE0, 0x00, 0x00, 0x01, 0x01, 0xC0, 0x61, 0x31, 0x31, 0x21, 0x20, 0xC0, 0x81,
    0x01, 0x01, 0x01, 0x00, 0x00, 0x00, 0x00, 0x01, 0x01, 0x01, 0x01, 0xFF, 0xFF,
    0x00, 0x00, 0x00, 0x00, 0x01, 0x03, 0x02, 0x03, 0x01, 0x00, 0x01, 0x03, 0xF2,
    0x1A, 0x0B, 0x08, 0x0B, 0x1B, 0x10, 0x60, 0xE3, 0x03, 0x00, 0x01, 0x03, 0x02,
    0x02, 0x03, 0x03, 0x00, 0x03, 0x03, 0x00, 0x00, 0x03, 0x03, 0x00, 0x00, 0x03,
    0x03, 0x00, 0x00, 0x03, 0x03, 0x03, 0x03, 0x00, 0x01, 0x03, 0x02, 0x02, 0x03,
    0x01, 0x00, 0x03, 0x03, 0x00, 0x00, 0x03, 0x00, 0x00, 0x00, 0x3E, 0x63, 0x80,
    0x80, 0x80, 0x80, 0x60, 0x3F, 0x07, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0xFE, 0x01, 0x01, 0x01, 0x02, 0x03, 0x3E, 0xE8, 0xF8, 0xF0, 0xD0, 0x90, 0x18,
    0x0F, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0xC0, 0x38, 0xFF, 0x0C, 0x38, 0xE0, 0x80, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xFF, 0xFF, 0x00,
    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x1F, 0xF0, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x33, 0x5F, 0x8F, 0x84, 0x05, 0x07, 0x06, 0x0C, 0x0E, 0x0E, 0x0C, 0x14, 0x34,
    0x68, 0x88, 0xD8, 0x70, 0x00, 0x00, 0x00, 0x00, 0x00, 0xE0, 0x10, 0x10, 0x10,
    0xF0, 0xE0, 0x00, 0xF0, 0xF0, 0x00, 0x80, 0x80, 0x00, 0x00, 0x80, 0x80, 0x80,
    0x80, 0x00, 0x80, 0x80, 0x00, 0x80, 0x00, 0x00, 0x20, 0x38, 0x0E, 0x01, 0xC0,
    0x3F, 0xE0, 0x00, 0x00, 0x03, 0x0E, 0x08, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
    0x00, 0x00, 0x00, 0xFF, 0xFF, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0xB6,
    0xED, 0xC0, 0xC0, 0xC0, 0xE0, 0xA0, 0xA0, 0xA0, 0xA0, 0xA1, 0xA1, 0xA1, 0xA1,
    0xA1, 0xA1, 0xA1, 0xE1, 0xE1, 0xC1, 0xEF, 0xBB, 0x83, 0x86, 0x88, 0xB0, 0x80,
    0x80, 0x80, 0x8F, 0x90, 0x90, 0x90, 0x9F, 0x8F, 0x80, 0x9F, 0x9F, 0x87, 0x8D,
    0x98, 0x80, 0x8C, 0x9E, 0x92, 0x92, 0x9F, 0xC0, 0xC7, 0xFF, 0xB8, 0x8F, 0x80,
    0x90, 0x90, 0xC0, 0xF0, 0x8E, 0x81, 0x80, 0x81, 0x8F, 0xB8, 0xE0, 0x80, 0x80,
    0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0x80, 0xFF,
);

my $rpi = Device::WebIO::RaspberryPi->new;
my $lcd = Device::PCD8544->new({
    dev      => 0,
    speed    => Device::PCD8544->SPEED_4MHZ,
    webio    => $rpi,
    power    => 2,
    rst      => 24,
    dc       => 23,
    contrast => 0x60,
    bias     => Device::PCD8544->BIAS_1_48,
});

$lcd->init;
$lcd->set_image( \@PIC_OUT );
$lcd->update;


lib/Device/PCD8544/Exceptions.pm
# Copyright (c) 2015  Timm Murray
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are met:
# 
#     * Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright 
#       notice, this list of conditions and the following disclaimer in the 
#       documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
package Device::PCD8544::Exceptions;

use v5.14;
use warnings;
use base 'Exception::Tiny';


package Device::PCD8544::ImageSizeException;

use base 'Device::PCD8544::Exceptions';


1;
__END__

=head1 NAME

  Device::PCD8544::Exceptions - All the exceptions we could throw

=cut


lib/Device/PCD8544/ConvertImage.pm
# Copyright (c) 2015  Timm Murray
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are met:
# 
#     * Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright 
#       notice, this list of conditions and the following disclaimer in the 
#       documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
package Device::PCD8544::ConvertImage;

use v5.14;
use warnings;
use Device::PCD8544::Exceptions;

use constant WIDTH  => 84;
use constant HEIGHT => 48;


sub convert
{
    my ($img) = @_;
    my $width = $img->getwidth;
    my $height = $img->getheight;
    if( ($width != WIDTH) || ($height != HEIGHT) ) {
        Device::PCD8544::ImageSizeException->throw( 'Width/height is'
            . " $width/$height"
            . ', expected is ' . WIDTH . '/' . HEIGHT
            . '.  Please rescale image.' );
    }

    my @lcd_bitmap = ();
    
    my $word = 0x00;
    my $pixels_counted = 0;
    foreach my $x (0 .. ($width-1)) {
        foreach my $y (0 .. ($height-1)) {
            my $pixel = $img->getpixel( x => $x, y => $y )
                or die "Could not get pixel ($x, $y)\n";

            my @channels = $pixel->rgba;
            my $val = $channels[0] ? 0 : 1;

            $word = ($word << 1) | $val;
            $pixels_counted++;

            if( $pixels_counted >= 8 ) {
                push @lcd_bitmap, $word;
                $word           = 0x00;
                $pixels_counted = 0;
            }
        }
    }
    return \@lcd_bitmap;
}


1;
__END__

=head1 NAME

  Device::PCD8544::ConvertImage - Convert an image to the format for the PCD8544 LCD

=cut


lib/Device/PCD8544.pm
# Copyright (c) 2015  Timm Murray
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without 
# modification, are permitted provided that the following conditions are met:
# 
#     * Redistributions of source code must retain the above copyright notice, 
#       this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright 
#       notice, this list of conditions and the following disclaimer in the 
#       documentation and/or other materials provided with the distribution.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE 
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE 
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
#
# Some of this code is based on Adafruit Industries' Python library, which is
# covered by:
#
#
# Copyright (c) 2014 Adafruit Industries
# Author: Tony DiCola
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.
#
package Device::PCD8544;

# ABSTRACT: Driver for the PCD8544 LCD controller (aka Nokia 5110)
use v5.14;
use warnings;
use Moo;
use namespace::clean;
use Device::WebIO::Device::DigitalOutput;
use Device::WebIO::Device::SPI;
use Time::HiRes ();

use constant DEBUG => 0;

use constant {
    BIAS_1_100 => 0b000,
    BIAS_1_80  => 0b001,
    BIAS_1_65  => 0b010,
    BIAS_1_48  => 0b011,
    BIAS_1_40  => 0b100,
    BIAS_1_24  => 0b101,
    BIAS_1_18  => 0b110,
    BIAS_1_10  => 0b111,
};
use constant {
    SPEED_1MHZ => 1_000_000,
    SPEED_2MHZ => 2_000_000,
    SPEED_4MHZ => 4_000_000,
};
use constant {
    POWERDOWN            => 0x04,
    ENTRYMODE            => 0x02,
    EXTENDED_INSTRUCTION => 0x01,
    DISPLAY_BLANK        => 0x00,
    DISPLAY_NORMAL       => 0x04,
    DISPLAY_ALL_ON       => 0x01,
    DISPLAY_INVERTED     => 0x05,
    FUNCTIONSET          => 0x20,
    DISPLAY_CONTROL      => 0x08,
    SETYADDR             => 0x40,
    SETXADDR             => 0x80,
    SETTEMP              => 0x04,
    SETBIAS              => 0x10,
    SETVOP               => 0x80,
};

has dev => (
    is       => 'ro',
    required => 1,
);
has speed => (
    is      => 'ro',
    default => sub { SPEED_4MHZ },
);
has webio => (
    is  => 'ro',
    isa => sub {
        my ($obj) = @_;
        if(! (
            $obj->does( 'Device::WebIO::Device::DigitalOutput' )
            && $obj->does( 'Device::WebIO::Device::SPI' )
        )) {
            die "The webio object must do the roles Device::WebIO::Device::DigitalOutput and Device::WebIO::Device::SPI\n";
        }
    },
    required => 1,
);
has 'power' => (
    is       => 'ro',
    required => 1,
);
has 'rst' => (
    is       => 'ro',
    required => 1,
);
has 'dc' => (
    is       => 'ro',
    required => 1,
);
has 'contrast' => (
    is      => 'rw',
    default => sub { 0x3C },
    trigger => sub {
        my ($self, $val) = @_;
        return 1 unless $self->_was_init_called;
        $self->_send_extended_command( SETVOP | $val );
        return 1;
    },
);
has 'bias' => (
    is      => 'rw',
    default => sub { BIAS_1_40 },
    trigger => sub {
        my ($self, $val) = @_;
        return 1 unless $self->_was_init_called;
        $self->_send_extended_command( SETBIAS | $val );
        return 1;
    },
);
has '_buffer' => (
    is      => 'rw',
    default => sub {[]},
);
has '_was_init_called' => (
    is      => 'rw',
    default => sub { 0 },
);


sub init
{
    my ($self) = @_;
    return 1 if $self->_was_init_called;
    my $webio = $self->webio;

    foreach ($self->power, $self->rst, $self->dc ) {
        $webio->set_as_output( $_ );
        $webio->output_pin( $_, 1 );
    }

    $webio->spi_set_speed( $self->dev, $self->speed );

    $self->reset;
    $self->_was_init_called( 1 );
    return 1;
}

sub reset
{
    my ($self) = @_;
    my $webio = $self->webio;
    my $power = $self->power;
    my $rst   = $self->rst;

    # Created using suggestions from 'Kuy' on Sparkfun product comments:
    # https://www.sparkfun.com/products/10168
    #
    $webio->output_pin( $power, 1 );
    $webio->output_pin( $rst,   1 );
    Time::HiRes::usleep( 5_000 );

    $webio->output_pin( $rst, 0 );
    Time::HiRes::usleep( 1_000 );

    $webio->output_pin( $rst, 1 );
    Time::HiRes::usleep( 5_000 );

    say "Setting bias (" . SETBIAS . " | " . $self->bias . ")" if DEBUG;
    $self->_send_extended_command( SETBIAS | $self->bias );
    say "Setting contrast (" . SETVOP . " | " . $self->contrast . ")" if DEBUG;
    $self->_send_extended_command( SETVOP  | $self->contrast );
    say "Setting Y Addr (" . SETYADDR . ")" if DEBUG;
    $self->_send_command( SETYADDR | 0x00 );
    say "Setting X Addr (" . SETXADDR . ")" if DEBUG;
    $self->_send_command( SETXADDR | 0x00 );

    return 1;
}

sub set_image
{
    my ($self, $img) = @_;
    $self->_buffer( $img );
    return 1;
}

sub update
{
    my ($self) = @_;
    $self->_send_buffer;
    return 1;
}

sub display_blank
{
    my ($self) = @_;
    $self->_send_command( DISPLAY_CONTROL | DISPLAY_BLANK );
    return 1;
}

sub display_normal
{
    my ($self) = @_;
    $self->_send_command( DISPLAY_CONTROL | DISPLAY_NORMAL );
    return 1;
}

sub display_all_on
{
    my ($self) = @_;
    $self->_send_command( DISPLAY_CONTROL | DISPLAY_ALL_ON );
    return 1;
}

sub display_inverse
{
    my ($self) = @_;
    $self->_send_command( DISPLAY_CONTROL | DISPLAY_INVERTED );
    return 1;
}

sub _send_command
{
    my ($self, $cmd) = @_;
    say "Sending command $cmd" if DEBUG;

    my $webio = $self->webio;
    $webio->output_pin( $self->dc, 0 );

    my $fmt_cmd = pack 'n', $cmd;
    $webio->spi_write( $self->dev, $fmt_cmd );

    return 1;
}

sub _send_extended_command
{
    my ($self, $cmd) = @_;
    say "Sending extended command $cmd {" if DEBUG;
    $self->_send_command( FUNCTIONSET | EXTENDED_INSTRUCTION );
    $self->_send_command( $cmd );
    $self->_send_command( FUNCTIONSET );
    $self->_send_command( DISPLAY_CONTROL | DISPLAY_NORMAL );
    say "}" if DEBUG;
    return 1;
}

sub _send_buffer
{
    my ($self) = @_;
    my $buffer = $self->_buffer;
    my $webio = $self->webio;

    $self->_send_command( SETYADDR  | 0x00 );
    $self->_send_command( SETXADDR | 0x00 );

    $webio->output_pin( $self->dc, 1 );

    my $fmt_buf = pack 'C*', @$buffer;
    $webio->spi_write( $self->dev, $fmt_buf );

    return 1;
}


1;
__END__

=head1 NAME

  Device::PCD8544 - Driver for the PCD8544 LCD controller (aka Nokia 5110)

=head1 SYNOPSIS

    my $rpi = Device::WebIO::RaspberryPi->new;
    my $lcd = Device::PCD8544->new({
        dev      => 0,
        speed    => Device::PCD8544->SPEED_4MHZ,
        webio    => $rpi,
        power    => 2,
        rst      => 24,
        dc       => 23,
        contrast => 0x3C,
        bias     => Device::PCD8544->BIAS_1_40,
    });
    
    $lcd->init;
    $lcd->set_image( \@PIC_OUT );
    $lcd->update;

=head1 DESCRIPTION

Implements the PCD8544 LCD controller using C<Device::WebIO>. This display is 
84x48 pixels of black-and-white.  It's similar to the one on the Nokia 5110 
phone.

=head1 METHODS

=head2 new

     new({
        dev      => 0,
        speed    => Device::PCD8544->SPEED_4MHZ,
        webio    => $webio,
        power    => 2,
        rst      => 24,
        dc       => 23,
        contrast => 0x60,
        bias     => Device::PCD8544->BIAS_1_48,
    }); 

Constructor.  Params:

=over 4

=item * dev: SPI device number against the C<webio> object

=item * speed: SPI speed to run at.  Using C<SPEED_4MHZ> is good for most purposes.  If you're running at much less than 3V, you may want to reduce this to C<SPEED_2MHZ>.

=item * webio: C<Device::WebIO> object that does the C<DigitalOutput> and C<SPI> roles.

=item * power: GPIO pin that's powering the device.  As the display itself only draws a few milliamps, it's safe to power it from GPIO.

=item * rst: GPIO pin for the reset wire.

=item * dc: GPIO pin for the Data/Control wire.

=item * contrast: Contrast level for the display.  This should be between 0x00 and 0x7F.  0x3C is usually a nice default.

=item * bias: Set bias level of the device.  Default is BIAS_1_48.  See the device datasheet for details.

=back

=head2 init

Sets up the pins and brings the device up.  Must be called before doing anything
else.

=head2 set_image

  set_image( \@IMG );

Sets an arrayref image.  Each entry is an 8-bit word, with each bit representing 
a pixel.  See C<Device::PCD8544::ConvertImage> for converting Imager objects 
into this format.

Be sure to call C<update()> to actually send the image to the display.

=head2 update

Send our image buffer to the display.

=head2 reset

Reset the device.

=head2 contrast

    contrast( 0x40 )

Set the contrast on the device.

=head2 bias

    $self->bias( $self->BIAS_1_80 )

Set the bias on the device.  See the datasheet for details.  The following 
constants are defined:

    BIAS_1_100
    BIAS_1_80
    BIAS_1_65
    BIAS_1_48
    BIAS_1_40
    BIAS_1_24
    BIAS_1_18
    BIAS_1_10

=head2 display_blank

Blank out all pixels on the display.

=head2 display_normal

Normal display mode of black on white background.

=head2 display_all_on

Turns on all pixels.

=head2 display_inverse

Inverse display mode of white on black background.

=head1 SEE ALSO

Datasheet: L<http://www.sparkfun.com/datasheets/LCD/Monochrome/Nokia5110.pdf>

=head1 LICENSE

Copyright (c) 2015  Timm Murray
All rights reserved.

Redistribution and use in source and binary forms, with or without modification, are 
permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright notice, this list of 
      conditions and the following disclaimer.
    * Redistributions in binary form must reproduce the above copyright notice, this list of
      conditions and the following disclaimer in the documentation and/or other materials 
      provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS 
OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF 
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE 
COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) 
HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR 
TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.


=cut
